pkg/build/patchbuild.go

```diff
@@ -0,0 +1,147 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package build
+
+import (
+	"bytes"
+	"fmt"
+	"text/template"
+
+	bundle "github.com/GoogleCloudPlatform/k8s-cluster-bundle/pkg/apis/bundle/v1alpha1"
+	"github.com/GoogleCloudPlatform/k8s-cluster-bundle/pkg/converter"
+	"github.com/GoogleCloudPlatform/k8s-cluster-bundle/pkg/filter"
+	"github.com/GoogleCloudPlatform/k8s-cluster-bundle/pkg/options"
+	"github.com/GoogleCloudPlatform/k8s-cluster-bundle/pkg/options/openapi"
+	"github.com/GoogleCloudPlatform/k8s-cluster-bundle/pkg/wrapper"
+	log "github.com/golang/glog"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// ComponentFiles reads file-references for component builder objects.
+// The returned components are copies with the file-references removed.
+func BuildPatchTemplate(ptb *bundle.PatchTemplateBuilder, opts options.JSONOptions) (*bundle.PatchTemplate, error) {
+	name := ptb.GetName()
+	if ptb.Template == "" {
+		return nil, fmt.Errorf("cannot build PatchTemplate from PatchTemplateBuilder %q: it has an empty template", name)
+	}
+
+	tmpl, err := template.New("ptb").Parse(ptb.Template)
+	if err != nil {
+		return nil, fmt.Errorf("cannot build PatchTemplate from PatchTemplateBuilder %q: error parsing template: %v", name, err)
+	}
+
+	if opts == nil {
+		opts = make(options.JSONOptions)
+	}
+
+	if ptb.BuildSchema != nil {
+		opts, err = openapi.ApplyDefaults(opts, ptb.BuildSchema)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// This is a hack to allow us to pass through runtime templates variables
```

Comment:
I haven't thought of a good non-hacky way to do this. I'm all ears =)

Reply:
Me either! But I realize that this ONLY works for simple substitutions. I can't do, for example

```
{{- if .ImagePullSecret != "" }}imagePullSecrets:
- name: {{.ImagePullSecret}}{{end}}
```

Because the "build" time template execution will execute the `if`.